storage/src/test/java/io/atomix/storage/journal/index/FileChannelJournalSegmentReaderTest.java

```diff
@@ -0,0 +1,75 @@
+package io.atomix.storage.journal.index;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import io.atomix.storage.StorageLevel;
+import io.atomix.storage.journal.Indexed;
+import io.atomix.storage.journal.JournalReader.Mode;
+import io.atomix.storage.journal.JournalSegmentDescriptor;
+import io.atomix.storage.journal.SegmentedJournal;
+import io.atomix.storage.journal.SegmentedJournalReader;
+import io.atomix.storage.journal.SegmentedJournalWriter;
+import io.atomix.utils.serializer.Namespace;
+import java.io.File;
+import java.io.IOException;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public class FileChannelJournalSegmentReaderTest {
+  private static final Namespace NAMESPACE = Namespace.builder().register(Integer.class).build();
+  private static final Integer ENTRY = 1;
+  private static final int ENTRY_SIZE =
+      NAMESPACE.serialize(ENTRY).length + Long.BYTES; // padding for checksum;
+
+  @Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  private File directory;
+
+  @Before
+  public void setUp() throws IOException {
+    directory = temporaryFolder.newFolder();
+  }
+
+  @Test
+  public void shouldResetBackwardsCorrectlyWhenUsingSameIndex() {
+    // given
+    final int entriesPerSegment = 7;
+    final SparseJournalIndex journalIndex = new SparseJournalIndex(5);
+    try (final SegmentedJournal<Integer> journal = createJournal(entriesPerSegment, journalIndex)) {
+      final SegmentedJournalWriter<Integer> writer = journal.writer();
+      final SegmentedJournalReader<Integer> reader = journal.openReader(1, Mode.ALL);
+
+      for (int i = 1; i <= entriesPerSegment; i++) {
+        writer.append(ENTRY);
+      }
+
+      writer.append(ENTRY);
+      final Indexed<Integer> previousEntry = writer.append(ENTRY);
+      final Indexed<Integer> currentEntry = writer.append(ENTRY);
+
+      reader.reset(currentEntry.index());
```

Comment:
So you told me this resets BEFORe this entry right?

Reply:
It resets it such that the next entry is the one at the given index. Also I just realized I put this file in the wrong package :see_no_evil:

Reply:
ups :man_shrugging: